plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/util/HudiAvroSerializer.java

```diff
@@ -207,13 +173,61 @@ else if (javaType == long.class) {
+                        // Directly get the float value from the Number
+                        // This preserves the fractional part
+                        float floatValue = ((Number) value).floatValue();
+
+                        // Get the IEEE 754 single-precision 32-bit representation of this float
+                        int intBits = Float.floatToRawIntBits(floatValue);
+
+                        // The writeLong method expects these int bits, passed as a long
+                        // NOTE: Java handles the widening conversion from int to long
+                        type.writeLong(output, intBits);
+                    }
+                    else {
+                        // Handle cases where 'value' is not a Number
+                        throw new TrinoException(GENERIC_INTERNAL_ERROR,
+                                format("Unhandled type for %s: %s | value type: %s", javaType.getSimpleName(), type, value.getClass().getName()));
+                    }
+                }
                 else if (type instanceof DecimalType decimalType) {
-                    verify(decimalType.isShort(), "The type should be short decimal");
-                    BigDecimal decimal = DECIMAL_CONVERTER.convert(decimalType.getPrecision(), decimalType.getScale(), value);
-                    type.writeLong(output, encodeShortScaledValue(decimal, decimalType.getScale()));
+                    if (value instanceof SqlDecimal sqlDecimal) {
+                        if (decimalType.isShort()) {
+                            writeShortDecimal(output, sqlDecimal.toBigDecimal().unscaledValue().longValue());
+                        }
+                        else {
+                            writeBigDecimal(decimalType, output, sqlDecimal.toBigDecimal());
+                        }
+                    }
+                    else if (value instanceof GenericData.Fixed fixed) {
+                        verify(decimalType.isShort(), "The type should be short decimal");
+                        BigDecimal decimal = DECIMAL_CONVERTER.convert(decimalType.getPrecision(), decimalType.getScale(), fixed.bytes());
+                        type.writeLong(output, encodeShortScaledValue(decimal, decimalType.getScale()));
+                    }
+                    else {
+                        throw new TrinoException(GENERIC_INTERNAL_ERROR,
+                                format("Unhandled type for %s: %s | value type: %s", javaType.getSimpleName(), type, value.getClass().getName()));
+                    }
                 }
                 else if (type.equals(DATE)) {
-                    type.writeLong(output, ((SqlDate) value).getDays());
+                    if (value instanceof SqlDate sqlDate) {
+                        type.writeLong(output, sqlDate.getDays());
+                    }
+                    else if (value instanceof Integer days) {
+                        ((DateType) type).writeInt(output, days);
+                    }
```

Comment:
For these logical types, is there any reference that how deserializer handles these types (as the serializer and deserializer has to be paired)?

Reply:
That's good enough for now.  Next step would be implementing `Page`-based reader context to plug into the file group reader for the Trino connector, so we can simplify the serialization and deserialization with improved performance wrt log files.